### ISSUE TITLE
Kobo Forma, CBZ, Ghosting: Even with full-refresh-every-page, midtone greys have ghosting after page-turn

### ISSUE THREAD

**Issue by angerson** [koreader/koreader#8880]
* KOReader version: 2022.02 (I don't know if this happens in way-earlier versions, but I first noticed it in 2021.12 before upgrading today)
* Device: Kobo Forma

#### Issue

Even though I have set Koreader to do a full refresh every page, turning the page on a comic causes ghostly artifacts to appear on mid-tone greys and gradients. This happens for both light-ghosts-on-darker backgrounds and dark-ghosts-on-lighter-backgrounds. Here is an example of what it looks like, after a page turn from a fabric checkerboard pattern to a simple gradient (source: [ghosting_example.cbz.zip](https://github.com/koreader/koreader/files/8201978/ghosting_example.cbz.zip)).

<details>
  <summary>photograph of my kobo forma screen after turning the page from pattern to gradient</summary>

![image](https://user-images.githubusercontent.com/32465472/157148983-cc196af1-8e77-4b41-ac74-8f16ad27fd0c.png)

</details>

- I dithered the image down to 8 colors to check and see if the effect was worse for a full greyscale color range vs a limited color set (since I found this with some preprocessed comic files with only 8 colors), but the effect seems to remain for any amount of colors.
- Doing a second full page refresh manually (without turning the page again) wipes all the artifacts, as expected, and looks great.
- If I set the Full Refresh Rate to "never" and flip pages a few times, the artifacts become pretty nasty, as expected. From that point, even if I do a "full refresh" with a gesture shortcut, there are still artifacts of the pattern shown on the gradient, and I have to do a second Full Refresh to make them truly disappear.

#### Steps to reproduce

- Open this file: [ghosting_example.cbz.zip](https://github.com/koreader/koreader/files/8201978/ghosting_example.cbz.zip)
- Set "Gear -> Screen -> E-ink Settings -> Full Refresh Rate -> Every Page"
- Turn the pages, observe the pattern is visible in the gradient even though the screen flashes black in between page turns
- Optionally, configure a shortcut to activate a full screen refresh. After a same-page refresh, the artifacts disappear, as expected, and the display looks great.

##### `crash.log` (if applicable)
Here's the relevant part of my crash.log from opening the example file, turning the pages a few times, and triggering a few manual full refreshes with a diagonal swipe shortcut.

[relevant.crash.log](https://github.com/koreader/koreader/files/8202099/relevant.crash.log)

**Comment by NiLuJe** [comment]
Nothing we can do (much). Fun fact: it's also affected by the amount of time the previous screen was shown. You'll notice that paging quickly will exhibit a different amount of ghosting ;o).

TL;DR: that's why we have a global shirt diagonal swipe gesture to force a refresh on demand ;).

**Comment by offset-torque** [comment]
@NiLuJe 

I will add this information to user guide to clarify the problem for users so they won't think this is caused by KOReader, their document or a defective screen. I too was under the impression that new generation screens can clear all "pixels" with a single refresh. Apparently they still can't.

Is this paragraph factually correct:

> Even when KOReader is set to full refresh after every page, you may still see some ghosting / residue after turning pages. This is a limitation of e-ink screen technology and can be observed even on the latest generation devices. As a workaround, you can refresh one more time manually to get rid of these artifacts if they bother you (short diagonal swipe gesture by default).

**Comment by Frenzie** [comment]
I suspect that's more likely to be only or primarily on the latest devices rather than even.

**Comment by NiLuJe** [comment]
It's generally limited to mixed content on recent devices, they've gotten fairly good at handling text ;).

**Comment by angerson** [comment]
That's a bummer. What if koreader had an option to refresh twice every page turn, equivalent to turning the page and then doing the full-refresh gesture -- would that still not work because of the time / temperature stuff?

**Comment by NiLuJe** [comment]
The timing aspect ensures that a double-refresh is not significantly better than a single one (it's possibly even worse), unless there's a somewhat significant pause between the two to let the screen's state "sit" for a while. And that's obviously a no-no ;).

Hence the gesture ;).

**Comment by angerson** [comment]
Makes sense. Damn e-ink magic!

What if Koreader was able to flash black and then white on a page turn, instead of just black -- a mega-refresh? I interspersed a white page between comic pages to test this, and ghosting decreased significantly. A built-in feature for black-then-white-then-page looks like it would improve comics with a lot of grays and gradients, although it would turn pages faster than my fingers so maybe it wouldn't be as effective.

**Comment by NiLuJe** [comment]
That's what the screensaver module does for images, and it's not perfect either (again, timing affects the end result a lot).
